src/generators/javascript/presets/CommonPreset.ts

```diff
@@ -0,0 +1,207 @@
+    const marshalCode = `json += \`"$\{key}": ${patternPropertyMarshalCode},\`;`;
+    marshalAdditionalProperties = `if(this.${additionalPropertyName} !== undefined) { 
+  for (const [key, value] of this.${additionalPropertyName}.entries()) {
+    //Only render additionalProperties which are not already a property
+    if(Object.keys(this).includes(String(key))) continue;
+    ${marshalCode}
+  }
+}`;
+  }
+  return marshalAdditionalProperties;
+}
+
+/**
+ * Render `marshal` function based on model
+ */
+function renderMarshal({ renderer, model, inputModel }: {
+  renderer: JavaScriptRenderer,
+  model: CommonModel,
+  inputModel: CommonInputModel
+}): string {
+  return `public marshal() : string {
+  let json = '{'
+${renderer.indent(renderMarshalProperties(model, renderer, inputModel))}
+${renderer.indent(renderMarshalPatternProperties(model, renderer, inputModel))}
+${renderer.indent(renderMarshalAdditionalProperties(model, renderer, inputModel))}
+
+  //Remove potential last comma 
+  return \`$\{json.charAt(json.length-1) === ',' ? json.slice(0, json.length-1) : json}}\`;
+}`;
+}
+
+function renderUnmarshalProperty(modelInstanceVariable: string, model: CommonModel, inputModel: CommonInputModel, renderer: JavaScriptRenderer) {
+  if (model.$ref) {
+    const resolvedModel = inputModel.models[model.$ref];
+    const propertyModelKind = TypeHelpers.extractKind(resolvedModel);
+    //Referenced enums only need standard marshalling, so lets filter those away
+    if (propertyModelKind !== ModelKind.ENUM) {
+      return `${renderer.nameType(model.$ref)}.unmarshal(${modelInstanceVariable})`;
+    }
+  }
+  return `${modelInstanceVariable}`;
+}
+function renderUnmarshalProperties(model: CommonModel, renderer: JavaScriptRenderer, inputModel: CommonInputModel) {
+  const properties = model.properties || {};
+  const propertyKeys = [...Object.entries(properties)];
+  const unmarshalProperties = propertyKeys.map(([prop, propModel]) => {
+    const formattedPropertyName = renderer.nameProperty(prop, propModel);
+    const modelInstanceVariable = `obj["${prop}"]`;
+    const unmarshalCode = renderUnmarshalProperty(modelInstanceVariable, propModel, inputModel, renderer);
+    return `if (${modelInstanceVariable} !== undefined) {
+  instance.${formattedPropertyName} = ${unmarshalCode};
+}`;
+  });
+  return unmarshalProperties.join('\n');
+}
+
+function renderUnmarshalPatternProperties(model: CommonModel, renderer: JavaScriptRenderer, inputModel: CommonInputModel) {
+  let unmarshalPatternProperties = '';
+  let setPatternPropertiesMap = '';
+  if (model.patternProperties !== undefined) {
+    for (const [pattern, patternModel] of Object.entries(model.patternProperties)) {
+      let patternPropertyName = getUniquePropertyName(model, `${pattern}${DefaultPropertyNames.patternProperties}`);
+      patternPropertyName = renderer.nameProperty(patternPropertyName, patternModel);
+      const modelInstanceVariable = 'value as any';
```

Comment:
```suggestion
      const modelInstanceVariable = 'value';
```
JS don't have type cast 🙂